pandas/core/arrays/period.py

```diff
@@ -85,94 +102,112 @@ def wrapper(self, other):
+
+        values = np.array(values, dtype='int64', copy=copy)
+        self._data = values
+        if freq is None:
+            raise ValueError('freq is not specified and cannot be inferred')
+        freq = Period._maybe_convert_freq(freq)
+        self._dtype = PeriodDtype(freq)
 
     @classmethod
     def _simple_new(cls, values, freq=None, **kwargs):
-        """
-        Values can be any type that can be coerced to Periods.
-        Ordinals in an ndarray are fastpath-ed to `_from_ordinals`
-        """
+        # alias from PeriodArray.__init__
+        return cls(values, freq=freq, **kwargs)
 
-        if is_period_dtype(values):
-            freq = dtl.validate_dtype_freq(values.dtype, freq)
-            values = values.asi8
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
+        # type: (Sequence[Optional[Period]], PeriodDtype, bool) -> PeriodArray
+        if dtype:
+            freq = dtype.freq
+        else:
+            freq = None
+        periods = np.asarray(scalars, dtype=object)
+        if copy:
+            periods = periods.copy()
 
-        if not is_integer_dtype(values):
-            values = np.array(values, copy=False)
-            if len(values) > 0 and is_float_dtype(values):
-                raise TypeError("{cls} can't take floats"
-                                .format(cls=cls.__name__))
-            return cls(values, freq=freq, **kwargs)
+        freq = freq or libperiod.extract_freq(periods)
+        ordinals = libperiod.extract_ordinals(periods, freq)
+        return cls(ordinals, freq=freq)
 
-        return cls._from_ordinals(values, freq=freq, **kwargs)
+    def _values_for_factorize(self):
+        return self.values, iNaT
 
     @classmethod
-    def _from_ordinals(cls, values, freq=None, **kwargs):
-        """
-        Values should be int ordinals
-        `__new__` & `_simple_new` cooerce to ordinals and call this method
-        """
-        # **kwargs are included so that the signature matches PeriodIndex,
-        #  letting us share _simple_new
+    def _from_factorized(cls, values, original):
+        # type: (Sequence[Optional[Period]], PeriodArray) -> PeriodArray
+        return cls(values, freq=original.freq)
 
-        values = np.array(values, dtype='int64', copy=False)
+    @classmethod
+    def _from_datetime64(cls, data, freq, tz=None):
+        """Construct a PeriodArray from a datetime64 array
 
-        result = object.__new__(cls)
-        result._data = values
-        if freq is None:
-            raise ValueError('freq is not specified and cannot be inferred')
-        result._freq = Period._maybe_convert_freq(freq)
-        return result
+        Parameters
+        ----------
+        data : ndarray[datetime64[ns], datetime64[ns, tz]]
+        freq : str or Tick
+        tz : tzinfo, optional
+
+        Returns
+        -------
+        PeriodArray[freq]
+        """
+        data = dt64arr_to_periodarr(data, freq, tz)
```

Comment:
can you assert the dtype here

Reply:
`dt64arr_to_periodarr` checks that first thing.